### ISSUE TITLE
MAINT: signal deprecation cleanups

### ISSUE THREAD

**Issue by tylerjereddy** [scipy/scipy#21724]
While reviewing gh-21706, I noticed some weird things about its precedent, gh-21211.

For starters, `correlate` performs the input validation on the same argument twice, and the testsuite still passes if I purge one of those duplicates:

```diff
--- a/scipy/signal/_signaltools.py
+++ b/scipy/signal/_signaltools.py
@@ -246,7 +246,6 @@ def correlate(in1, in2, mode='full', method='auto'):
     in1 = np.asarray(in1)
     in2 = np.asarray(in2)
     _reject_objects(in1, 'correlate')
-    _reject_objects(in1, 'correlate')
```

Curiously, the testsuite also still passes if I "hide" the `object` dtype in the second (unvalidated) array:

```diff
--- a/scipy/signal/tests/test_signaltools.py
+++ b/scipy/signal/tests/test_signaltools.py
@@ -2101,7 +2101,7 @@ class TestCorrelate:
 
     def test_dtype_deprecation(self):
         # gh-21211
-        a = np.asarray([1, 2, 3, 6, 5, 3], dtype=object)
+        a = np.asarray([1, 2, 3, 6, 5, 3], dtype=np.int32)
         b = np.asarray([2, 3, 4, 5, 3, 4, 2, 2, 1], dtype=object)
         with pytest.deprecated_call(match="dtype=object is not supported"):
             correlate(a, b)
```

this appears to be the case because `convolve` masquerades as `correlate` for the purposes of its deprecation messages (and `convolve` gets called after the `in2` object sneaks past its intended gate):

```python
import numpy as np
import scipy
from scipy.signal import convolve


convolve(np.arange(5).astype(object),
         np.ones(5).astype(object))
```

```
/private/tmp/test.py:6: DeprecationWarning: dtype=object is not supported by correlate and will raise an error in SciPy 1.17.0. Supported dtypes are: boolean, integer, `np.float16`,`np.float32`, `np.float64`, `np.complex64`, `np.complex128`.
  convolve(np.arange(5).astype(object),
```

**Comment by ev-br** [comment]
Great catch Tyler! gh-21713 has a fix:  https://github.com/scipy/scipy/pull/21713/files#diff-351836adc98d076c1552d17a57c52e6aa8ca43760bae44bea9190e55b4769b7fR250

(not that the fix was deliberate---the whole `_reject_objects` call needed a bit of care for non-numpy backends)

**Comment by lucascolley** [comment]
@ev-br perhaps you could split the fix out of that PR to make it into 1.15.0? (I don't think I'll get round to reviewing the PR in time, sorry!)

**Comment by ev-br** [comment]
Okay, https://github.com/scipy/scipy/pull/21984.
